core/include/gpio.h

```diff
@@ -23,16 +26,42 @@ enum gpio_interrupt {
 
 struct gpio_chip {
 	const struct gpio_ops *ops;
+	unsigned int gpio_base;
+	unsigned int ngpios;
+	vaddr_t base;
+
+	SLIST_ENTRY(gpio_chip) link;
 };
 
 struct gpio_ops {
-	enum gpio_dir (*get_direction)(unsigned int gpio_pin);
-	void (*set_direction)(unsigned int gpio_pin, enum gpio_dir direction);
-	enum gpio_level (*get_value)(unsigned int gpio_pin);
-	void (*set_value)(unsigned int gpio_pin, enum gpio_level value);
-	enum gpio_interrupt (*get_interrupt)(unsigned int gpio_pin);
-	void (*set_interrupt)(unsigned int gpio_pin,
-		enum gpio_interrupt ena_dis);
+	enum gpio_dir (*get_direction)(struct gpio_chip *gc, unsigned int pin);
+	void (*set_direction)(struct gpio_chip *gc, unsigned int pin,
+			      enum gpio_dir direction);
+	enum gpio_level (*get_value)(struct gpio_chip *gc, unsigned int pin);
+	void (*set_value)(struct gpio_chip *gc, unsigned int pin,
+			  enum gpio_level value);
+	enum gpio_interrupt (*get_interrupt)(struct gpio_chip *gc,
+					     unsigned int pin);
+	void (*set_interrupt)(struct gpio_chip *gc, unsigned int pin,
+			      enum gpio_interrupt ena_dis);
+};
+
+struct gpio_desc {
+	unsigned int pin;
+	struct gpio_chip *gc;
+	void *dev;
+
+	SLIST_ENTRY(gpio_desc) link;
```

Comment:
Thanks for the suggestion. I felt its heavy on op-tee light weight model. But my design is based on following. I am happy to adapt a more lighter approach.

Only the used GPIO will add a descriptor to the gdlist which should be really very few in secure world.  And once a consumer of pin gets hold of the  *gd  it can directly refer the ops to work on
instead of having to lookup for which *gc (chip/controller) ops to use during each operation.

This structure can be extended to hold the interrupt handler for the pin.

```suggestion
struct gpio_desc {
struct gpio_chip *gc;
uint32_t  (*consumer_itr_handler)(void *dev);  
void *dev;   /* this can be something else.. like the *pdata*/
SLIST_ENTRY(gpio_desc) link;
}

struct gpio_chip {
       const struct gpio_ops *ops;
       unsigned int gpio_base;
       struct itr_handler *gc_handler;
       SLIST_ENTRY(gpio_desc) link;
}
```
gc_handler.handler() - for each pending bit/pin in gc  -> get_desc(pin) -> gd->consumer_itr_handler(gd->dev)
apart from interrupt handling it can also ensure resource /pin sharing as gd->dev  is  kind of reserving the pin by the consumer. I have not used the spin lock yet, but that can be added.

Reply:
> Only the used GPIO will add a descriptor to the gdlist which should be really very few in secure world

I think this assumption is false. This GPIO framework is intended to be generic and maybe a platforms will need to handle many GPIOs. Yet I understand this need for GPIO event notification when the is not a specific interrupt lines for GPIOs in a GPIO group.

>  I have not used the spin lock yet, but that can be added.

If a GPIO chip needs spinlocking before accessing resource interface, then it should be handled from the chip driver, not from the GPIO framework.

This GPIO framework needs spinlocking, but only locally, to protect modifications in common data as the chip and descriptor lists.